python/hail/methods/statgen.py

```diff
@@ -2908,15 +2909,103 @@ def filter(self):
+    is a local pruning step, which prunes variants in the same partition. The parallelism
+    in this step will be affected by the number of partitions in the dataset, as well as
+    the memory per core, which is used to calculate a queue size for the local prune.
+
+    The second stage is a global pruning step which makes use of a correlation matrix
+    across all remaining variants. In the global pruning step, correlated variants are
+    passed to a method that computes a maximal independent set of those variants.
+
+    Parameters
+    ----------
+    ds : :class:`.MatrixTable`
+        dataset to prune
+    r2 : :obj:`float`
+        correlation threshold (exclusive) above which variants are removed
+    window : :obj:`int`
+        distance in kilobases; correlation between variants further apart is not considered in pruning
+    memory_per_core : :obj:`int`
+        memory in MB per core for local pruning
+
+    Returns
+    -------
+    :class:`.Table`
+        Table of variants after pruning.
+    """
+    bytes_per_core = memory_per_core * 1024 * 1024
+    fraction_memory_to_use = 0.25
+    variant_byte_overhead = 50
+    genotypes_per_pack = 32
+    n_samples = ds.count_cols()
+    min_memory_per_core = math.ceil((1 / fraction_memory_to_use) * 8 * n_samples + variant_byte_overhead)
+    if bytes_per_core < min_memory_per_core:
+        raise ValueError("memory per core must be greater than {} MB".format(min_memory_per_core * 1024 * 1024))
+    bytes_per_variant = math.ceil(8 * n_samples / genotypes_per_pack) + variant_byte_overhead
+    memory_available_per_core = bytes_per_core * fraction_memory_to_use
+    max_queue_size = int(max(1.0, math.ceil(memory_available_per_core / bytes_per_variant)))
+
+    sites_only_table = Table(Env.hail().methods.LocalLDPrune.apply(
+        require_biallelic(ds, 'ld_prune')._jvds, float(r2), window, max_queue_size))
+
+    sites_path = new_temp_file()
+    sites_only_table.write(sites_path, overwrite=True)
+    sites_only_table = hl.read_table(sites_path)
+
+    locally_pruned_ds = ds.annotate_rows(pruned=sites_only_table[ds.row_key])
+    locally_pruned_ds = locally_pruned_ds.filter_rows(hl.is_defined(locally_pruned_ds.pruned))
+
+    locally_pruned_ds = (locally_pruned_ds.annotate_rows(
+        mean=locally_pruned_ds.pruned.mean, sd_reciprocal=locally_pruned_ds.pruned.sd_reciprocal)
+        .add_row_index('row_idx'))
+
+    locally_pruned_path = new_temp_file()
+    locally_pruned_ds.write(locally_pruned_path, overwrite=True)
+    locally_pruned_ds = hl.read_matrix_table(locally_pruned_path)
+
+    normalized_mean_imputed_genotype_expr = (
+        hl.cond(hl.is_defined(locally_pruned_ds.GT),
+                (locally_pruned_ds.GT.n_alt_alleles() - locally_pruned_ds.mean)
+                * locally_pruned_ds.sd_reciprocal, 0))
+
+    # BlockMatrix.from_entry_expr writes to disk
+    block_matrix = BlockMatrix.from_entry_expr(normalized_mean_imputed_genotype_expr, block_size=512)
```

Comment:
The advantage of smaller block size for dropping a bit more area from the BlockMatrix is likely outweighed by the additional overhead per partition in the multiply. Consider leaving the default block size until you have benchmark results to show smaller is better.

Reply:
All times are an average of 3 tests. The time to compute the correlation matrix and entries table must be measured together to take advantage of laziness of the matrix multiply.

12,000 variants

| Block Size | 128 | 256 | 512 | 1024 | 2048 | 4096 |
| ---------- | ---- | ---- | --- | ---- | ---- | ---- |
| Time to compute Correlation Matrix and Entries Table | 69sec | 61 | **56** | 61 | 67 | 78 |
| Total Time for LD Prune | 233 | 197 | 174 | **169** | 182 | 189 |

~26k variants

| Block Size | 256 | 512 | 1024 | 2048 | 4096 |
| ---------- | ---- | --- | ---- | ---- | ---- |
| Time to compute Correlation Matrix and Entries Table | 149sec | **148** | 163 | 244 | 226 | 
| Total Time for LD Prune | 392 | 354 | **348** | 406 | 409 | 

I'm not sure if the positional density of the variants in datasets varies much from dataset to dataset, in which case the ideal block size for other datasets might be different than the ideal block size for the datasets I tested here. If that were the case, it might make sense to make block size a parameter. For now, I changed it to 1024.